homeassistant/components/template/event.py

```diff
@@ -0,0 +1,270 @@
+        try:
+            self._trigger_event(event_type, {})
+        except ValueError:
+            _LOGGER.error(
+                "Received invalid event_type: %s for entity %s. Expected one of: %s",
+                event_type,
+                self.entity_id,
+                self._attr_event_types,
+            )
+
+
+class TemplateEvent(TemplateEntity, AbstractTemplateEvent):
+    """Representation of a template event."""
+
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config: dict[str, Any],
+        unique_id: str | None,
+    ) -> None:
+        """Initialize the select."""
+        TemplateEntity.__init__(self, hass, config=config, unique_id=unique_id)
+        AbstractTemplateEvent.__init__(self, config)
+        if TYPE_CHECKING:
+            assert self._attr_name is not None
+
+        self._attr_device_info = async_device_info_to_link_from_device_id(
+            hass, config.get(CONF_DEVICE_ID)
+        )
+
+    @callback
+    def _async_setup_templates(self) -> None:
+        """Set up templates."""
+        self.add_template_attribute(
+            "_attr_event_types",
+            self._event_types_template,
+            None,
+            self._update_event_types,
+            none_on_template_error=True,
+        )
+        self.add_template_attribute(
+            "_event_type",
+            self._event_type_template,
+            None,
+            self._update_event_type,
+            none_on_template_error=True,
+        )
+        super()._async_setup_templates()
+
+
+class TriggerEventEntity(TriggerEntity, AbstractTemplateEvent, RestoreEntity):
+    """Event entity based on trigger data."""
+
+    domain = EVENT_DOMAIN
+    extra_template_keys_complex = (
+        CONF_EVENT_TYPE,
+        CONF_EVENT_TYPES,
+    )
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        coordinator: TriggerUpdateCoordinator,
+        config: ConfigType,
+    ) -> None:
+        """Initialize the entity."""
+        TriggerEntity.__init__(self, hass, coordinator, config)
+        AbstractTemplateEvent.__init__(self, config)
+
+        # Render the _attr_name before initializing TriggerEventEntity
+        self._attr_name = self._rendered.get(CONF_NAME, DEFAULT_NAME)
+
+        self._attr_device_info = async_device_info_to_link_from_device_id(
+            hass, config.get(CONF_DEVICE_ID)
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle update of the data."""
+        self._process_data()
+
+        if not self.available:
+            self.async_write_ha_state()
+            return
+
+        for key, updater in (
+            (CONF_EVENT_TYPES, self._update_event_types),
+            (CONF_EVENT_TYPE, self._update_event_type),
+        ):
+            updater(self._rendered[key])
+
+        self.async_set_context(self.coordinator.data["context"])
+        self.async_write_ha_state()
```

Comment:
Doesn't this mean we always set the context before writing state?

I never worked with context so no clue how that works

Reply:
Yes, for trigger based template entities, we update the context from whatever triggered.  Kinda like automations.  There was a "bug fix" awhile ago that added this because template entities didn't properly reflect the context according to some people.